File: python/p21/p21.py
# http://projecteuler.net/problem=21

import math


def divisors(n):
    """ Returns the divisors of n that are less than n """
    dvs = []
    for i in range(1, int(math.sqrt(n)) + 1):
        if n % i == 0:
            dvs.append(i)
            j = n / i
            if j != i:
                dvs.append(j)

    dvs.remove(n)
    return dvs

def d(n):
    """ Finds the sum of the divisors less than n """
    return sum(divisors(n))


# Compute the list of amicable numbers less than 10,000 and print their sum
amicable_nums = set()
for a in range(2, 10000):
    b = d(a)
    if d(b) == a and a != b:
        amicable_nums.add(a)
        amicable_nums.add(b)

print sum(amicable_nums)

File: csharp/P21_to_30/P21.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using ProjectEuler.Helpers;

namespace ProjectEuler.P21_to_30
{
    /// <summary>
    /// https://projecteuler.net/problem=21
    /// </summary>
    internal class P21
    {
        private const int Max = 10000;

        /// <summary>
        /// Calculates the sum of all amicable numbers under 10,000
        /// </summary>
        public int Solve()
        {
            var total = 0;

            for (var i = 1; i < Max; i++)
            {
                var sum = SumOfDivisors(i);
                if (sum != i && SumOfDivisors(sum) == i)
                {
                    // Found amicable pair; add i to total
                    total += i;
                }
            }

            return total;
        }

        /// <summary>
        /// Returns the sum of the proper divisors of a given number.
        /// Proper divisors are all numbers less than n that evenly divide n.
        /// </summary>
        private int SumOfDivisors(int n)
        {
            // Exclude n as a divisor since it is not a proper divisor
            return PrimeHelpers.Divisors(n).Sum() - n;
        }
    }
}